### DISCUSSION TITLE
Getting values below an object below an event

### DISCUSSION THREAD

**Discussion by gobidani** [FreeOpcUa/opcua-asyncio#1195] [Q&A]
Hi!

I would like to access some variables that are defined under an object. But the object is under an event.

Example server:

```python
import asyncio
import logging
from asyncua import ua
from asyncua.server import Server, EventGenerator

logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger('asyncua')

async def main():
    server = Server()
    await server.init()
    server.set_endpoint("opc.tcp://0.0.0.0:4840/freeopcua/server/")
    uri = "http://examples.freeopcua.github.io"
    idx = await server.register_namespace(uri)
    
    myobj = await server.nodes.objects.add_object(idx, "MyObject")

    custom_etype = await server.nodes.base_event_type.add_object_type(idx, 'MyEvent')
    eventInt = await custom_etype.add_property(idx, 'MyIntProperty', ua.Variant(0, ua.VariantType.Int64))
    eventObj = await custom_etype.add_object(idx, 'MyEventObject')
    eventVar = await eventObj.add_variable(idx, "MyEventVariable", 10) # This is what I'm looking for
    myevgen = await server.get_event_generator(custom_etype, myobj)

    async with server:
        count = 0
        while True:
            await asyncio.sleep(1)
            new_val = await eventInt.get_value() + 1
            await eventInt.write_value(new_val % 1024)
            new_val = await eventVar.get_value() + 1
            await eventVar.write_value(new_val % 1024)
            count = (count + 1) % 1024
            await myevgen.trigger(message="MyEvent %d" % count)

if __name__ == "__main__":
    asyncio.run(main())
```

Example client:

```python
import asyncio
import logging
from asyncua import Client
from asyncua.common.events import get_filter_from_event_type, get_event_obj_from_type_node

logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger(__name__)

class SubHandler:
    def event_notification(self, event):
        print(event) # This is where I want to receive it

async def main():
    url = "opc.tcp://localhost:4840/freeopcua/server/"
    async with Client(url=url) as client:
        obj = await client.nodes.root.get_child(["0:Objects", "2:MyObject"])
        myevent = await client.nodes.root.get_child(["0:Types", "0:EventTypes", "0:BaseEventType", "2:MyEvent"])
        myEventObject = await client.nodes.root.get_child(["0:Types", "0:EventTypes", "0:BaseEventType", "2:MyEvent", "2:MyEventObject"])
        myEventObjectFilter = await get_filter_from_event_type([myEventObject]) # This is not working, server log shows event does not fit where clause

        msclt = SubHandler()
        sub = await client.create_subscription(100, msclt)
        handle = await sub.subscribe_events(obj, [myevent], myEventObjectFilter)

        await asyncio.sleep(4)
        await sub.unsubscribe(handle)
        await sub.delete()

if __name__ == "__main__":
    asyncio.run(main())
```

Without the filter, I only get the property directly below the event (MyIntProperty), but I want the object (MyEventObject) and its variable (MyEventVariable).
Of course the server could be wrong, I'm mostly interested in the client side and simulate the server.
I'm new to OPC UA, so my understanding could also be wrong.
Can someone show me an example of how to get an object and its variables / properties below an event?

Thank you in advance.

**Comment by AndreasHeine** [comment]
you need to modify your server!

```python
    myobj = await server.nodes.objects.add_object(idx, "MyObject")
    custom_etype = await server.nodes.base_event_type.add_object_type(2, 'MySecondEvent')
    await custom_etype.add_property(2, 'MyIntProperty', ua.Variant(0, ua.VariantType.Int32))
    await custom_etype.add_property(2, 'MyBoolProperty', ua.Variant(True, ua.VariantType.Boolean))
    mysecondevgen = await server.get_event_generator(custom_etype, myobj)
```

create an own eventtype with added custom properties (like a copy of your objecttype with only properies and variables) and if you set the props and vars you can emit the event containing the data you are looking for ;)

as a reference look into the opcua types and you will see that many eventtypes are copies from objecttypes
so lets say whenever a event occour you will get the data from that point in time of that object

**Reply by AndreasHeine** [reply]
![image](https://user-images.githubusercontent.com/56362817/217000710-aa43511f-7e76-446f-af40-417169840794.png)

**Reply by gobidani** [reply]
Maybe there is a misunderstanding, maybe in my head about OPC UA events :)
So I have a given server, see image:

![Server](https://user-images.githubusercontent.com/124591128/217004235-75d566a2-f99c-4508-8495-17a640da60d5.png)

The green part is the root node, the black part is an event.
The event contains some basic properties, e.g. **JobCycleCounter**.
And it contains the **User** object, with e.g. the **Name** property.

In the client, I want to access the **User**/**Name** property when the **UserFeedbackLog** event is triggered on the server.

I tried to simulate a similar server and use event filter without success.

**Reply by gobidani** [reply]
I think I figured out a workaround. It works with a test server, which is hopefully correct.
So I use ```subscribe_data_change``` to subscribe to **UserFeedbackLog/User/Name** to keep track of its changes.
And I use ```subscribe_events``` without filters to subscribe to the event **UserFeedbackLog**.
When the event triggers, I have the value of the variable I need already stored in my program.
